File: email.py
class Change:
    @staticmethod
    def get_all_emails(email_changes):
        user_emails = {}
        visited = set()

        for current_email, prior_email in email_changes.items():
            if current_email in visited:
                continue
            
            visited.add(current_email)
            emails = user_emails.get(current_email, [])
            
            if prior_email is not None:
                # traverse prior email recursively
                emails.extend(Change.traverse_prior_emails(prior_email, email_changes, visited))
            
            emails.append(current_email)
            user_emails[current_email] = emails
        
        return user_emails

    @staticmethod
    def traverse_prior_emails(prior_email, email_changes, visited):
        if prior_email in visited:
            return
            
        visited.add(prior_email)

        email = email_changes.get(prior_email)
        emails = []

        if email is not None:
            emails.extend(Change.traverse_prior_emails(email, email_changes, visited))
        
        emails.append(prior_email)
        return emails

if __name__ == "__main__":
    email_changes = {
        "a@example.com": "b@example.com",
        "b@example.com": "c@example.com",
        "x@example.com": "y@example.com",
        "y@example.com": "z@example.com",
        # Add more entries if needed
    }

    all_emails = Change.get_all_emails(email_changes)

    print(all_emails)

File: email.java
import java.util.*;

public class Change {
    public static Map<String, List<String>> getAllEmails(Map<String, String> emailChanges) {
        Map<String, List<String>> userEmails = new HashMap<>();
        Set<String> visited = new HashSet<String>();

        for (Map.Entry<String, String> entry : emailChanges.entrySet()) {
            String currentEmail = entry.getKey();
            if (visited.contains(currentEmail)) {
                continue;
            }
            if (!visited.contains(currentEmail)){
                visited.add(currentEmail);
            }
            List<String> emails = userEmails.getOrDefault(currentEmail, new ArrayList<>());

            String priorEmail = entry.getValue();
            if (priorEmail != null) {
                // traverse prior email recursively
                emails.addAll(traversePriorEmails(priorEmail, emailChanges));
            }

            emails.add(currentEmail);
            userEmails.put(currentEmail, emails);
        }

        return userEmails;
    }

    private static List<String> traversePriorEmails(String priorEmail, Map<String, String> emailChanges) {
        String email = emailChanges.get(priorEmail);
        List<String> emails = new ArrayList<>();

        if (email != null) {
            emails.addAll(traversePriorEmails(email, emailChanges));
        }

        emails.add(priorEmail);
        return emails;
    }

    public static void main(String[] args) {
        Map<String, String> emailChanges = new HashMap<>();
        emailChanges.put("a@example.com", "b@example.com");
        emailChanges.put("b@example.com", "c@example.com");
        emailChanges.put("x@example.com", "y@example.com");
        emailChanges.put("y@example.com", "z@example.com");
        // Add more entries if needed

        Map<String, List<String>> allEmails = getAllEmails(emailChanges);
        
        System.out.println(allEmails);
    }
}
